### ISSUE TITLE
Disappearing constant values

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1806]
Migrated from [rt.perl.org#3066](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3066) (status was 'resolved')

Searchable as RT3066$

**Comment by p5pRT** [comment]
### From eric@xmission.com

#### Created by eblood@xmission.com



When useithreads is defined\, the following code does not generate
a 'Modification of a read\-only value attempted' error like it does
in perl5\.005\_03\.

&nbsp;  sub C \(\) \{ 1 \}
&nbsp;  sub M \{ $\_\[0\] = 2; \}
&nbsp;  eval "C";
&nbsp;  M\(C\);

The following code shows a much more serious consequence of the
same bug\.

&nbsp;  sub C \(\) \{ 1 \}
&nbsp;  sub M \{ M\(\) if defined\(C\); \}
&nbsp;  eval "C";
&nbsp;  M\(C\);

This code should go into infinite recursion\.  However in
perl5\.6\.0 using ithreads it drops out of recursion after
the second iteration because C becomes undefined\.

This is caused by a number of interacting pieces of code as follows&#8203;:

&nbsp;  1\. When the tokenizer encounters a bareword whose definition
&nbsp;     is a constant sub\, it gets the SV for the constant in the
&nbsp;     sub and puts it in the op\_sv field of the opcode
&nbsp;     \(after incrementing the SV's reference count\)\.
&nbsp;     As a result multiple opcodes may all point to the same SV\.
&nbsp;  2\. Peep moves this SV to the pad\.  When it does so it sets the
&nbsp;     SVs\_PADTMP flag\.  At this point multiple pad slots may all
&nbsp;     be pointing to the same SV\.
&nbsp;  3\. Opcodes sometimes get cleared before the end of the program
&nbsp;     \(when they are were generated by eval or by the main body
&nbsp;     of a module\)\.  When the opcode is cleared it calls pad\_free
&nbsp;     on it's op\_targ value\.
&nbsp;  4\. pad\_free clears the SVs\_PADTMP and SVs\_READONLY flags on the
&nbsp;     SV\.  Unfortunately this SV may also be used in other pads\.
&nbsp;  5\. When a procedure recurses it copies the pad constants but
&nbsp;     only if the associated SV is read\-only\.  Since the read\-only
&nbsp;     flag was cleared by pad\_free the constant doesn't get copied
&nbsp;     and appears as undef in the recursive call\.

I can see several possible solutions but I'm not sure which
approach to take\.  I'm willing to write a patch but I'd like
some advice on which way to go\.  Or if someone with more experience
in the perl core feels they should do this instead that's fine with
me\.

These are the possibilities I see&#8203;:

&nbsp;  1\. Have the tokenizer always create a new SV for constant subs\.
&nbsp;  2\. Have peek copy the constant instead of using it directly when
&nbsp;     it moves it to the pad\.
&nbsp;  3\. Have pad\_free never remove the flags when both SVs\_PADTMP and
&nbsp;     SVs\_READONLY are set\.
&nbsp;  4\. Have pad\_free only remove the flags when the reference count
&nbsp;     for the SV is 1\.

I lean towards 4 or 2\.  I think that 4 would be safe but it might
not always free up a pad slot when it should\.  2 should also be
safe but would cause additional memory usage\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.0:

Configured by eblood at Fri Mar 24 12:36:06 2000.

Summary of my perl5 (revision 5 version 6 subversion 0) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cl', optimize='-O1 -MD -DNDEBUG', gccversion=
    cppflags='-DWIN32'
    ccflags ='-O1 -MD -DNDEBUG -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DPERL_MSVCRT_READFIX'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release  -libpath:"D:\apps\perl56\lib\CORE"  -machine:x86'
    libpth="g:\apps\msdevstd\lib" "g:\apps\msdevstd\mfc\lib" "%lib%" "D:\apps\perl56\lib\CORE"
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl56.lib
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release  -libpath:"D:\apps\perl56\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.0:
    d:\apps\cmt\lib
    D:/apps/perl56/lib
    D:/apps/perl56/site/lib
    .


Environment for perl v5.6.0:
    HOME=d:\home
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=d:\apps\perl56\bin;D:\WINNT\system32;D:\WINNT;g:\apps\bor\Delphi5\Projects\Bpl;g:\apps\bor\Delphi5\Bin;d:\apps\perl\bin;d:\apps\cmt\nt;d:\bin\nt;d:\bin;;g:\apps\bor\delphi3\bin;g:\apps\bor\tasm\bin;g:\apps\bor\bp\bin;d:\apps\iqtools;g:\apps\msdevstd\bin\;g:\apps\bor\bcc55\bin;;d:\rnd;c:\APPS\PERL\BIN
    PERLLIB=d:\apps\cmt\lib
    PERL_BADLANG (unset)
    PERL_MAILERS=smtp
    SHELL (unset)


```
</details>

**Comment by p5pRT** [comment]
### From @gsar

On Sun\, 09 Apr 2000 22&#8203;:21&#8203;:27 MDT\, eric@&#8203;xmission\.com wrote&#8203;:
>\[excellent analysis snipped\]
>
>These are the possibilities I see&#8203;:
>
>  1\. Have the tokenizer always create a new SV for constant subs\.
>  2\. Have peek copy the constant instead of using it directly when
>     it moves it to the pad\.
>  3\. Have pad\_free never remove the flags when both SVs\_PADTMP and
>     SVs\_READONLY are set\.
>  4\. Have pad\_free only remove the flags when the reference count
>     for the SV is 1\.
>
>I lean towards 4 or 2\.  I think that 4 would be safe but it might
>not always free up a pad slot when it should\.  2 should also be
>safe but would cause additional memory usage\.

I'd say peek\(\) should make a copy when moving it to the pad if the
SV is already a PADTMP\.

Thanks\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:
>
>I'd say peek\(\) should make a copy when moving it to the pad if the
>SV is already a PADTMP\.

What does PADTMP \_mean\_ ?  \(And where does one look fror such definitions?\)

All I can find is comment in sv\.h
"
sv\.h&#8203;:\#define SVs\_PADTMP    0x00000200    /\* in use as tmp \*/
"

Is setting PADTMP on a constant the right thing to do in the first place?

**Comment by p5pRT** [comment]
### From @gsar

On Mon\, 10 Apr 2000 08&#8203;:26&#8203;:14 BST\, Nick Ing\-Simmons wrote&#8203;:
>Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:
>>
>>I'd say peek\(\) should make a copy when moving it to the pad if the
>>SV is already a PADTMP\.
>
>What does PADTMP \_mean\_ ?  \(And where does one look fror such definitions?\)

It simply signifies that the SV is lives on the scratchpad\.  Such SVs
are of two kinds\.  One kind has names \(i\.e\. have corresponding entries
in the comppadname array\) and are usually lexical/our variables\.  The
other kind are nameless "temporaries" that may be allocated by any OP
for use during its lifetime\.

>Is setting PADTMP on a constant the right thing to do in the first place?

Yes\.  The constant SV is moved to the pad in order to keep it separate
between different threads that might share the same op tree\.  \(Otherwise\,
you'd have to lock access to the constant\, which we don't want to do\.
And besides\, constants are subject sv\_upgrade\(\) etc\.\, which will break
things when multiple threads share the same SV\.\)


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:
>On Mon\, 10 Apr 2000 08&#8203;:26&#8203;:14 BST\, Nick Ing\-Simmons wrote&#8203;:
>>Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:
>>>
>>>I'd say peek\(\) should make a copy when moving it to the pad if the
>>>SV is already a PADTMP\.
>>
>>What does PADTMP \_mean\_ ?  \(And where does one look fror such definitions?\)
>
>It simply signifies that the SV is lives on the scratchpad\.  Such SVs
>are of two kinds\.  One kind has names \(i\.e\. have corresponding entries
>in the comppadname array\) and are usually lexical/our variables\.  The
>other kind are nameless "temporaries" that may be allocated by any OP
>for use during its lifetime\.

So a "better" name would be "IN\_A\_PAD" ;\-\)

>
>>Is setting PADTMP on a constant the right thing to do in the first place?
>
>Yes\.  The constant SV is moved to the pad in order to keep it separate
>between different threads that might share the same op tree\.  

Presumably then this is one of the 'named' kind \- as we don't want
constants being used as scratch\. 
So perhaps there are three types&#8203;:
&nbsp;  1\. named variables
&nbsp;  2\. constants 
&nbsp;  3\. Pad temporaries\. 

>\(Otherwise\,
>you'd have to lock access to the constant\, which we don't want to do\.
>And besides\, constants are subject sv\_upgrade\(\) etc\.\, which will break
>things when multiple threads share the same SV\.\)

Ick\. I thought you could not upgrade if SvREADONLY was on?
I have certainly been bitten by that in the past\.

**Comment by p5pRT** [comment]
### From @gsar

On Mon\, 10 Apr 2000 17&#8203;:53&#8203;:38 BST\, Nick Ing\-Simmons wrote&#8203;:
>Gurusamy Sarathy \<gsar@&#8203;ActiveState\.com> writes&#8203;:
>>>Is setting PADTMP on a constant the right thing to do in the first place?
>>
>>Yes\.  The constant SV is moved to the pad in order to keep it separate
>>between different threads that might share the same op tree\.  
>
>Presumably then this is one of the 'named' kind \- as we don't want
>constants being used as scratch\. 

"Used as scratch" depends on the particular op\.  While that is
usually the case\, it is not always so\.  \(It is probably more
approprate to think of it as a "holding cell" rather than as
a "scratch value"\.\)

>So perhaps there are three types&#8203;:
>  1\. named variables
>  2\. constants 
>  3\. Pad temporaries\. 

There's also&#8203;:

&nbsp;   4\. GVs

Note that \#2 and \#4 are only applicable to ithreads\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Monday\, April 10\, 2000 12&#8203;:48 AM\, "Gurusamy Sarathy" 
\<gsar@&#8203;ActiveState\.com> wrote&#8203;:
> On Sun\, 09 Apr 2000 22&#8203;:21&#8203;:27 MDT\, eric@&#8203;xmission\.com wrote&#8203;:
> >\[excellent analysis snipped\]
> >
> >These are the possibilities I see&#8203;:
> >
> >  1\. Have the tokenizer always create a new SV for constant subs\.
> >  2\. Have peek copy the constant instead of using it directly when
> >     it moves it to the pad\.
> >  3\. Have pad\_free never remove the flags when both SVs\_PADTMP and
> >     SVs\_READONLY are set\.
> >  4\. Have pad\_free only remove the flags when the reference count
> >     for the SV is 1\.
> >
> >I lean towards 4 or 2\.  I think that 4 would be safe but it might
> >not always free up a pad slot when it should\.  2 should also be
> >safe but would cause additional memory usage\.
> 
> I'd say peek\(\) should make a copy when moving it to the pad if the
> SV is already a PADTMP\.
> 

Sounds good\.  Here's the patch\.

<details><summary>Inline Patch</summary>

```diff
--- op.c~ Fri Mar 24 22:24:40 2000
+++ op.c Mon Apr 10 21:17:16 2000
@@ -6417,10 +6417,17 @@
       * for reference counts, sv_upgrade() etc. */
      if (cSVOP->op_sv) {
   PADOFFSET ix = pad_alloc(OP_CONST, SVs_PADTMP);
-  SvREFCNT_dec(PL_curpad[ix]);
-  SvPADTMP_on(cSVOPo->op_sv);
-  PL_curpad[ix] = cSVOPo->op_sv;
-  cSVOPo->op_sv = Nullsv;
+  if (SvPADTMP(cSVOPo->op_sv)) {
+    /* If op_sv is already a PADTMP then it is being used by */
+    /* another pad so make a copy.                           */
+    sv_setsv(PL_curpad[ix],cSVOPo->op_sv);
+    SvREADONLY_on(PL_curpad[ix]);
+  } else {
+    SvREFCNT_dec(PL_curpad[ix]);
+    SvPADTMP_on(cSVOPo->op_sv);
+    PL_curpad[ix] = cSVOPo->op_sv;
+    cSVOPo->op_sv = Nullsv;
+  }
   o->op_targ = ix;
      }
 #endif
```
</details>

**Comment by p5pRT** [comment]
### From @gsar

On Mon\, 10 Apr 2000 21&#8203;:27&#8203;:58 MDT\, "Eric Blood" wrote&#8203;:
>Sounds good\.  Here's the patch\.
>
>\-\-\- op\.c~ Fri Mar 24 22&#8203;:24&#8203;:40 2000
>\+\+\+ op\.c Mon Apr 10 21&#8203;:17&#8203;:16 2000
>@&#8203;@&#8203; \-6417\,10 \+6417\,17 @&#8203;@&#8203;
>       \* for reference counts\, sv\_upgrade\(\) etc\. \*/
>      if \(cSVOP\->op\_sv\) \{
>   PADOFFSET ix = pad\_alloc\(OP\_CONST\, SVs\_PADTMP\);
>\-  SvREFCNT\_dec\(PL\_curpad\[ix\]\);
>\-  SvPADTMP\_on\(cSVOPo\->op\_sv\);
>\-  PL\_curpad\[ix\] = cSVOPo\->op\_sv;
>\-  cSVOPo\->op\_sv = Nullsv;
>\+  if \(SvPADTMP\(cSVOPo\->op\_sv\)\) \{
>\+    /\* If op\_sv is already a PADTMP then it is being used by \*/
>\+    /\* another pad so make a copy\.                           \*/
>\+    sv\_setsv\(PL\_curpad\[ix\]\,cSVOPo\->op\_sv\);
>\+    SvREADONLY\_on\(PL\_curpad\[ix\]\);

I haven't looked closely\, but I'd think we need a couple of lines
here to avoid a scalar leak&#8203;:

&nbsp;      SvREFCNT\_dec\(cSVOPo\->op\_sv\);
&nbsp;      cSVOPo\->op\_sv = Nullsv;

>\+  \} else \{
>\+    SvREFCNT\_dec\(PL\_curpad\[ix\]\);
>\+    SvPADTMP\_on\(cSVOPo\->op\_sv\);
>\+    PL\_curpad\[ix\] = cSVOPo\->op\_sv;
>\+    cSVOPo\->op\_sv = Nullsv;
>\+  \}
>   o\->op\_targ = ix;
>      \}
> \#endif

Thanks for the patch\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Monday\, April 10\, 2000 9&#8203;:58 PM\, "Gurusamy Sarathy" 
\<gsar@&#8203;ActiveState\.com> wrote&#8203;:

> On Mon\, 10 Apr 2000 21&#8203;:27&#8203;:58 MDT\, "Eric Blood" wrote&#8203;:
> 
> I haven't looked closely\, but I'd think we need a couple of lines
> here to avoid a scalar leak&#8203;:
> 
>       SvREFCNT\_dec\(cSVOPo\->op\_sv\);
>       cSVOPo\->op\_sv = Nullsv;
> 

Eeek\!\! You're right we do need those lines but I don't think its
because we would have scalar leak \(the reference count should be
decremented when the op is cleared which is admittedly later than
it needs to be\)\.  However\, the real problem is that pp\_const 
will use op\_sv instead of the pad value if op\_sv is set\.  So we
could have multiple threads all accessing the same constant SV\.
Not good\. 

Here's a \(hopefully\) corrected patch\.

<details><summary>Inline Patch</summary>

```diff
--- op.c~ Fri Mar 24 22:24:40 2000
+++ op.c Mon Apr 10 22:21:50 2000
@@ -6417,10 +6417,19 @@
       * for reference counts, sv_upgrade() etc. */
      if (cSVOP->op_sv) {
   PADOFFSET ix = pad_alloc(OP_CONST, SVs_PADTMP);
-  SvREFCNT_dec(PL_curpad[ix]);
-  SvPADTMP_on(cSVOPo->op_sv);
-  PL_curpad[ix] = cSVOPo->op_sv;
-  cSVOPo->op_sv = Nullsv;
+  if (SvPADTMP(cSVOPo->op_sv)) {
+    /* If op_sv is already a PADTMP then it is being used by */
+    /* another pad so make a copy.                           */
+    sv_setsv(PL_curpad[ix],cSVOPo->op_sv);
+    SvREADONLY_on(PL_curpad[ix]);
+    SvREFCNT_dec(cSVOPo->op_sv);
+    cSVOPo->op_sv = Nullsv;
+  } else {
+    SvREFCNT_dec(PL_curpad[ix]);
+    SvPADTMP_on(cSVOPo->op_sv);
+    PL_curpad[ix] = cSVOPo->op_sv;
+    cSVOPo->op_sv = Nullsv;
+  }
   o->op_targ = ix;
      }
 #endif
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed in 6a7129a1f\, aka change \#5979\.
